### DISCUSSION TITLE
Fields dependending on previous choices

### DISCUSSION THREAD

**Discussion by anthonygalea** [mhuebert/inside-out#8] [Q&A] [answered]
Thank you for creating inside-out 🙏 

## Context
We are running into an issue around validation in some forms. A small example to reproduce the problem:
```clj
(let [r (r/atom :option-1)]
  (with-form [f (cond-> {:a ?a
                         :b ?b}
                  (= :option-1 @r)
                  (merge {:c ?c
                          :d ?d}))]

    [component-a]
    [component-b]
    [component-r
     {:value @r
      :on-change #(reset! r %)}]
    (when (= :option-1 @r)
      [component-c]
      [component-d])

    [component-submit
     {:disabled? (forms/invalid? f)
      :payload @f}]))
```

Derefing `f` behaves as expected i.e. we have a map with `:a` and `:b` and if the user chooses `:option-1` we also get `:c` and `:d`. So far so good.

## Problem
The validation is performed on all 4 fields (irrespective on whether `:option-1` is chosen or not). This means that if the user chooses `:option-1`, modifies field `?c` (such as to make it invalid), and then switches away from `:option-1`, `(forms/invalid? f)` still returns `true` (even though field `?c` is no longer in "scope"). 

If I followed correctly the chain is: `invalid? -> valid? -> messages -> descendants -> !children` and `!children` is always returning all fields.

## Alternatives considered
1. use `forms/add-to-parent!` and `forms/remove-binding!` accordingly when the r/atom changes
2. replace the call to `forms/invalid?` with:

```clj
(let [irrelevant-fields #{'?c '?d}]
  (seq (remove #(contains? irrelevant-fields (:sym %))
               (forms/messages form :deep true))))
```

Is there some alternative way of solving this?

**Comment by mhuebert** [comment, marked as answer]
Hi @anthonygalea, that's a good question. I might handle this by having the "sub"-field check if it is required by reading from the 2nd arg (context / sibling fields):

```clj
:validators {?c (fn [value siblings]
                  (when (= :option-1 @('?r siblings))
                    (when-not (#{:X :Y :Z} value)
                      "Value must be :X, :Y or :Z")))}
```

I added an example to the notebook for this case: https://github.com/mhuebert/inside-out/blob/1ea1c041698a128ec713afb7d6a5784a15a39cbb/dev/inside_out/notebook.clj#L544

**Reply by anthonygalea** [reply]
Thanks @mhuebert. A colleague had actually suggested that and I discarded it as an option because `r` was not a field but a ratom. The reason I chose to make `r` a ratom was because I was under the (wrong) assumption that only things that end up in the payload can be inside-out fields. Now I realize that:
1. even if it were to end up in the payload (as in the example in your notebook), it wouldn't matter anyway because the backend ignores it
2. my assumption was wrong, it is possible to use a field without including it in the payload:

```clojure
(with-form [f (cond-> {:a ?a
                       :b ?b}
                (= :option-1 ?r)
                (merge {:c ?c
                        :d ?d}))
            :init {?r :some-option}]
  ...)
```

Thanks again
